### ISSUE TITLE
[NOT AN ISSUE] Event View for calendar

### ISSUE THREAD

**Issue by ApoorveTyagi** [patchthecode/JTAppleCalendar#920]
This is not an issue but a query.
Is there a way to add event views on the calendar other than the dots.
I wanted to create something like multi-day events just like you have on iPad calendar (screenshot attached
![screenshot](https://user-images.githubusercontent.com/10254070/46912589-2c987380-cf97-11e8-98d1-7400fad7bb94.png)
).
I tired by adding a view on date cell but don't like that approach as single event is divided into multiple views. I wanted a single view to be drawn over date cells.

**Comment by patchthecode** [comment]
I am not sure why you didnt like that approach.
Your event is spread across multiple dates. ok i got that.
But that shouldnt have anything to do with the calendar/cells.

What does your data source look like for your multiple events?

**Comment by ApoorveTyagi** [comment]
My datasource is a collection of events. With each event having its start and end date.
for ex. { 
"name" : "Event1",
"startDate" : "Oct 14, 2018 6:00:00 AM",
"endDate" : "Oct 20, 2018 6:00:00AM"
}

Here as you can see, this single event spans over multiple days. How do you suggest I do this? It should look like this:

![simulator screen shot - ipad air 2 - 2018-10-14 at 21 15 38](https://user-images.githubusercontent.com/10254070/46918992-51700380-cff6-11e8-8a0d-438a5ef523ff.png)

**Comment by patchthecode** [comment]
The library is just a UICollectionView.
Therefore you just have to make your datasource something similar to be compatible with a UICollectionView.

In a normal UICollectionView, each cell is mapped to an indexPath, but in this library every cell is mapped to a date.

Therefore, why are you saving your datasource as a range, when the library which is supposed to consume the datasource is expecting single-dates and **not** a range like you currently have?

The fix would be to change your datasource to be compatible with the library which has to consume it.
Do not have your datasource saved as a range of dates. Have it saved as single dates.
Something similar this will do (psuedo code):

var dataSource: [Date: Event] = [
   date1: Event1,
   date2:Event1,
   date3:Event1,
   date4: Event2 // etc etc
]

**Comment by ApoorveTyagi** [comment]
Thank you for taking the time out to explain this.
That's the same approach I am currently using. However, there are a few drawbacks with this approach in regards to the exact UI that I want.
for ex: I have to write the title "Event1" exactly in the middle of that blue event bar, which won't be possible with this approach as the exact middle might lie between two dates.
There are many more small UI things that I am not able to achieve with this approach.

That is why I had a second approach in mind: 
I create a one more data source function that takes in events.
Each section of collection view will now have "number of cells" + "number of events" items.
Then I update JTAppleCalendarLayout so that it first draws cells and then events independently.

Would this be fine or will it become too complex?

**Comment by patchthecode** [comment]
its not complex.
but i have one question before i think on this, lets say the events take up 8 cells (one full row and one cell on the other row), then where do you want the centering label to be?

**Comment by ApoorveTyagi** [comment]
I"ll always calculate the center as :  
center of label = (total length of event view) / 2
In this case, total length of event view = 8 * cellWidth.

One more use case, if the event is 14 days long.
In that case, the center will lie between 7th and 8th cell. Here I can decide, either to always have the label on end of 1st row or start of 2nd row.

**Comment by patchthecode** [comment]
it looks like you need to setup some sort of overlayview which are not managed by the cells (if i got you correctly). I say this because you want a label's text to lie between to cells ( in some cases ). Im not sure what is the best way to go about this. I have never had a UICollectionView where i needed to have a view that lay between cells before.

Let me know ont he mode you are thinking of and maybe i can have more ideas on this.

**Comment by MamtaKukreja** [comment]
@ApoorveTyagi I have the same requirement as you. How were you able to solve this?

**Comment by patchthecode** [comment]
@MamtaKukreja 
He didn't respond last time. But what I would say is --> This lib was not built to do something like this.  

However, if any developer has accomplished this with a regular UICollectionView (not this calendar), then please let me know how it was done. That way i can look into possible implementing this.

**Comment by ApoorveTyagi** [comment]
@MamtaKukreja , @patchthecode

I was able to accomplish this by modifying the JTAppleCalendar itself.
Although mine wasn’t an elegant solution but I was able to draw the events over the cells.

Here's what I did:

- Collectionview by default supports UICollectionElementKindSectionHeader and UICollectionElementKindSectionFooter types of supplementary views. I created a third type: **UICollectionElementKindEvent**.

-  Overwrote 
public func collectionView(_ collectionView: UICollectionView, **viewForSupplementaryElementOfKind** kind: String, at indexPath: IndexPath) -> UICollectionReusableView {
	if kind == UICollectionElementKindEvent {
            return calendarDelegate!.calendar(collectionView as! JTAppleCalendarView, eventViewForEventAt: indexPath)
        }...

here **eventViewForEventAt** was a delegate method that I created, where I would pass a reusable view to be shown. This view is subclass of JTAppleCollectionReusableView.

- I also created one more datasoruce method **func getEventsAt(section: Int) -> [EventConfig]** to pass my eventconfiguration to calendarview. **EventConfig** has a start and an end Date.

- Now in JTAppleCalendarLayout class, we layout our cells/decorationviews/supplementaryview in configureHorizontalLayout() & configureVerticalLayout(). As my calendar was only vertical, i just overwrote configureVerticalLayout(), got the event configuration for each section from above datasource method and calculated the position for my UICollectionElementKindEvent views. Calculating positions took time as an event could span over multiple rows.

I hope this gives you some direction as I won't be able to share any code.
